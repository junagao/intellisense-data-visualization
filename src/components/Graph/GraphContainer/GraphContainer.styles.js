import styled from 'styled-components';

export const GraphDiv = styled.div``;

export const GraphSvg = styled.svg`
  background-color: #f8f9fa;
  padding: 1rem;
`;
