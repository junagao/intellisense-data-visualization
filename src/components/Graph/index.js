export { default as Axis } from './Axis';
export { default as AxisLabel } from './AxisLabel';
export { default as GraphContainer } from './GraphContainer';
export { default as Lines } from './Lines';
export { default as LinesPoints } from './LinesPoints';
export { default as Tooltip } from './Tooltip';
