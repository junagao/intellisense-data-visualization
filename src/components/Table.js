import React from 'react';
import PropTypes from 'prop-types';

const Table = ({ data }) => {
  const renderTableHeader = () => {
    if (data.length) {
      const header = Object.keys(data[0]);
      return header
        .slice(0, 2)
        .map((key) => (
          <th key={key}>{key.charAt(0).toUpperCase() + key.slice(1)}</th>
        ));
    }
    return null;
  };

  const renderTableData = () =>
    data.map((item) => {
      const { metric, value } = item;
      return value && value[0] !== 'OK' ? (
        <tr key={metric}>
          <td>{metric}</td>
          <td>{value[value.length - 1].toFixed(2)}</td>
        </tr>
      ) : null;
    });

  return (
    <div>
      <table>
        <thead>
          <tr>{renderTableHeader()}</tr>
        </thead>
        <tbody>{renderTableData()}</tbody>
      </table>
    </div>
  );
};

Table.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      metric: PropTypes.string,
      value: PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      ),
    }),
  ).isRequired,
};

export default Table;
