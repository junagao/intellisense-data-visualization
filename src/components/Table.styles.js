import styled from 'styled-components';

export const TableContainer = styled.div`
  min-width: 80%;
`;

export const TableTable = styled.table`
  width: 100%;
  font-size: 0.9rem;
  text-align: left;
  border-radius: 2px 2px 0 0;
  border-collapse: separate;
  border-spacing: 0;
`;

export const TableThead = styled.thead`
  display: table-header-group;
  vertical-align: middle;
  border-top-left-radius: 2px;
  border-top-right-radius: 2px;
`;

export const TableTh = styled.th`
  color: rgba(0, 0, 0, 0.85);
  background-color: #fafafa;
  border-bottom: 1px solid #f0f0f0;
  display: table-cell;
  font-weight: 500;
  text-align: left;
  transition: background 0.3s ease;
  padding: 1rem;
`;

export const TableTr = styled.tr`
  display: table-row;
  position: relative;
  vertical-align: middle;
  &:hover {
    background-color: rgba(0, 0, 0, 0.05);
  }
`;

export const TableTd = styled.td`
  padding: 0.75rem 1rem;
  color: rgba(0, 0, 0, 0.65);
`;
