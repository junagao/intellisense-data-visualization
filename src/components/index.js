export { default as Error } from 'components/Error';
export { default as Loading } from 'components/Loading';
export { default as Table } from 'components/Table';
export {
  Axis,
  AxisLabel,
  GraphContainer,
  Lines,
  LinesPoints,
  Tooltip,
} from 'components/Graph';
