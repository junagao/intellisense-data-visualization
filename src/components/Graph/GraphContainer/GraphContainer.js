import React from 'react';
import PropTypes from 'prop-types';
import * as d3 from 'd3';

import { Axis, AxisLabel, Lines, Tooltip } from 'components';
import { GraphDiv, GraphSvg } from './GraphContainer.styles';

const chartSettings = {
  width: 650,
  height: 450,
  marginTop: 30,
  marginRight: 5,
  marginBottom: 45,
  marginLeft: 80,
  lineColors: [
    '#4EAF63',
    '#EC624F',
    '#DBC36C',
    '#F369A3',
    '#FDB462',
    '#CCEBC5',
    '#B3DE69',
    '#E04530',
    '#8DD3C7',
    '#8BC0E4',
    '#5676DC',
    '#5664C7',
    '#8952A5',
    '#BC80BD',
    '#BDB4DE',
    '#FCCDE5',
    '#FFED6F',
    '#949494',
  ],
};

const GraphContainer = ({
  data,
  selectedMetrics,
  onMouseMove,
  onHover,
  legendXPosition,
  legendYPosition,
  hoveredMetricColor,
  hoverLineGraph,
  hoveredMetric,
  hoveredTime,
  hoveredValue,
}) => {
  if (!data && !selectedMetrics) return {};

  const {
    width,
    height,
    marginTop,
    marginRight,
    marginBottom,
    marginLeft,
    lineColors,
  } = chartSettings;

  const dataToDraw = selectedMetrics.length ? selectedMetrics : data;

  const xMin = d3.min(dataToDraw, ({ dataset }) =>
    d3.min(dataset, ({ time }) => time),
  );

  const xMax = d3.max(dataToDraw, ({ dataset }) =>
    d3.max(dataset, ({ time }) => time),
  );

  const yMax = d3.max(dataToDraw, ({ dataset }) =>
    d3.max(dataset, ({ value }) => value),
  );

  const xScale = d3
    .scaleLinear()
    .domain([xMin, xMax])
    .range([marginLeft, width - marginRight]);

  const yScale = d3
    .scaleLinear()
    .domain([0, yMax])
    .range([height - marginBottom, marginTop]);

  const metricNames = dataToDraw.map(({ metric }) => metric);

  const color = d3.scaleOrdinal().domain(metricNames).range(lineColors);

  const metrics = dataToDraw.map((item) => {
    const line = d3
      .line()
      .x((d) => xScale(d.time))
      .y((d) => yScale(d.value))
      .curve(d3.curveNatural)(item.dataset);

    return {
      metric: item.metric,
      path: line,
      color: color(item.metric),
      dataset: item.dataset,
    };
  });

  return (
    <GraphDiv>
      {hoverLineGraph ? (
        <Tooltip
          x={legendXPosition}
          y={legendYPosition}
          color={hoveredMetricColor}
          metric={hoveredMetric}
          time={hoveredTime}
          value={hoveredValue}
        />
      ) : null}
      <GraphSvg
        width={width + marginLeft + marginRight}
        height={height + marginBottom}
        onMouseMove={onMouseMove}
      >
        <g transform={`translate(${marginLeft}, 0)`}>
          <Axis
            scales={{ xScale, yScale }}
            margin={{ marginTop, marginRight, marginBottom, marginLeft }}
            svgDimensions={{ width, height }}
          />
          <AxisLabel
            x={-height / 2}
            y={marginRight - marginLeft}
            dy="1rem"
            transform="rotate(-90)"
            label="Values"
          />
          <AxisLabel
            x={width / 2 - marginLeft - marginRight}
            y={height + 5}
            label="Times"
          />
          <Lines
            metrics={metrics}
            transform={`translate(-${marginLeft}, 0)`}
            onHover={onHover}
          />
          {metrics.map(({ metric, dataset, color }) => (
            <g key={color} fill={color}>
              {dataset.map(({ time, value }) => (
                <circle
                  key={time}
                  cx={xScale(time) - marginLeft}
                  cy={yScale(value)}
                  r="3"
                  stroke="#fff"
                  onMouseOver={() => onHover(metric, color, time, value)}
                  onFocus={() => onHover(metric, color, time, value)}
                  onMouseOut={() => onHover(metric, color, time, value)}
                  onBlur={() => onHover(metric, color, time, value)}
                />
              ))}
            </g>
          ))}
        </g>
      </GraphSvg>
    </GraphDiv>
  );
};

GraphContainer.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      metric: PropTypes.string,
      dataset: PropTypes.arrayOf(
        PropTypes.shape({
          time: PropTypes.number,
          value: PropTypes.number,
        }),
      ),
    }),
  ).isRequired,
  selectedMetrics: PropTypes.arrayOf(
    PropTypes.shape({
      metric: PropTypes.string,
      dataset: PropTypes.arrayOf(
        PropTypes.shape({
          time: PropTypes.number,
          value: PropTypes.number,
        }),
      ),
    }),
  ).isRequired,
  onMouseMove: PropTypes.func.isRequired,
  onHover: PropTypes.func.isRequired,
  legendXPosition: PropTypes.number.isRequired,
  legendYPosition: PropTypes.number.isRequired,
  hoveredMetricColor: PropTypes.string.isRequired,
  hoverLineGraph: PropTypes.bool.isRequired,
  hoveredMetric: PropTypes.string.isRequired,
  hoveredTime: PropTypes.string.isRequired,
  hoveredValue: PropTypes.string.isRequired,
};

export default GraphContainer;
