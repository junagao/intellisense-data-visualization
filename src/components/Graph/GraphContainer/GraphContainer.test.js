import React from 'react';
import { shallow } from 'enzyme';

import { Tooltip, Axis, AxisLabel, Lines } from 'components';
import GraphContainer from './GraphContainer';

describe('GraphContainer', () => {
  let wrapper;

  const mockData = [
    {
      metric: 'metric A',
      dataset: [
        { time: 0, value: 10 },
        { time: 5, value: 39 },
        { time: 10, value: 50 },
        { time: 15, value: 130 },
      ],
    },
    {
      metric: 'metric B',
      dataset: [
        { time: 0, value: 0 },
        { time: 5, value: 24 },
        { time: 10, value: 43 },
        { time: 15, value: 390 },
      ],
    },
  ];
  const mockSelectedMetrics = [];

  beforeEach(() => {
    wrapper = shallow(
      <GraphContainer
        data={mockData}
        selectedMetrics={mockSelectedMetrics}
        onMouseMove={jest.fn()}
        onHover={jest.fn()}
        legendXPosition={100}
        legendYPosition={200}
        hoveredMetricColor="#ff0"
        hoverLineGraph
        hoveredMetric="metric A"
        hoveredTime="20"
        hoveredValue="134"
      />,
    );
  });

  it('should render correctly given the required props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render one GraphDiv element', () => {
    expect(wrapper.find('GraphDiv').length).toEqual(1);
  });

  it('should render one instance of Tooltip if hover on line graph', () => {
    expect(wrapper.find(Tooltip).length).toEqual(1);
  });

  it('should render null hoverLineGraph false', () => {
    wrapper.setProps({ hoverLineGraph: false });
    expect(wrapper.find(Tooltip).length).toEqual(0);
  });

  it('should render one GraphSvg element', () => {
    expect(wrapper.find('GraphSvg').length).toEqual(1);
  });

  it('should render one g elements', () => {
    expect(wrapper.find('g').length).toEqual(1);
  });

  it('should render one instance of Axis', () => {
    expect(wrapper.find(Axis).length).toEqual(1);
  });

  it('should render two instances of AxisLabel', () => {
    expect(wrapper.find(AxisLabel).length).toEqual(2);
  });

  it('should render one instance of Lines', () => {
    expect(wrapper.find(Lines).length).toEqual(1);
  });
});
