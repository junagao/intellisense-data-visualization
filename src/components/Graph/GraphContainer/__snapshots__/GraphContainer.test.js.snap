// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphContainer should render correctly given the required props 1`] = `
<GraphDiv>
  <Tooltip
    color="#ff0"
    metric="metric A"
    time="20"
    value="134"
    x={100}
    y={200}
  />
  <GraphSvg
    height={495}
    onMouseMove={[MockFunction]}
    width={735}
  >
    <g
      transform="translate(80, 0)"
    >
      <Axis
        margin={
          Object {
            "marginBottom": 45,
            "marginLeft": 80,
            "marginRight": 5,
            "marginTop": 30,
          }
        }
        scales={
          Object {
            "xScale": [Function],
            "yScale": [Function],
          }
        }
        svgDimensions={
          Object {
            "height": 450,
            "width": 650,
          }
        }
      />
      <AxisLabel
        dy="1rem"
        label="Values"
        transform="rotate(-90)"
        x={-225}
        y={-75}
      />
      <AxisLabel
        dy={null}
        label="Times"
        transform={null}
        x={240}
        y={455}
      />
      <Lines
        marginLeft={80}
        metrics={
          Array [
            Object {
              "color": "#4EAF63",
              "dataset": Array [
                Object {
                  "time": 0,
                  "value": 10,
                },
                Object {
                  "time": 5,
                  "value": 39,
                },
                Object {
                  "time": 10,
                  "value": 50,
                },
                Object {
                  "time": 15,
                  "value": 130,
                },
              ],
              "metric": "metric A",
              "path": "M80,395.38461538461536C142.77777777777777,383.076923076923,205.55555555555554,370.7692307692308,268.33333333333337,367.5C331.1111111111112,364.2307692307692,393.8888888888889,370,456.6666666666667,356.92307692307696C519.4444444444445,343.8461538461539,582.2222222222222,311.923076923077,645,280.00000000000006",
            },
            Object {
              "color": "#EC624F",
              "dataset": Array [
                Object {
                  "time": 0,
                  "value": 0,
                },
                Object {
                  "time": 5,
                  "value": 24,
                },
                Object {
                  "time": 10,
                  "value": 43,
                },
                Object {
                  "time": 15,
                  "value": 390,
                },
              ],
              "metric": "metric B",
              "path": "M80,405C142.77777777777777,389.87179487179486,205.55555555555554,374.74358974358984,268.33333333333337,381.92307692307696C331.1111111111112,389.1025641025641,393.8888888888889,418.5897435897436,456.6666666666667,363.65384615384613C519.4444444444445,308.7179487179487,582.2222222222222,169.35897435897434,645,30",
            },
          ]
        }
        onHover={[MockFunction]}
        scales={
          Object {
            "xScale": [Function],
            "yScale": [Function],
          }
        }
        transform="translate(-80, 0)"
      />
    </g>
  </GraphSvg>
</GraphDiv>
`;
